File: utbot-spring-analyzer/src/main/java/analyzer/UtBotBeanFactoryPostProcessor.java
package analyzer;

import org.springframework.beans.BeansException;
import org.springframework.beans.factory.annotation.AnnotatedBeanDefinition;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.beans.factory.support.BeanDefinitionRegistry;
import org.springframework.core.PriorityOrdered;
import org.springframework.core.type.MethodMetadata;

import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.Arrays;

public class UtBotBeanFactoryPostProcessor implements BeanFactoryPostProcessor, PriorityOrdered {

    @Override
    public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {

        System.out.println("Started post-processing bean factory in UtBot");

        ArrayList<String> beanClassNames = new ArrayList<>();
        for (String beanDefinitionName : beanFactory.getBeanDefinitionNames()) {
            BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanDefinitionName);

            if (beanDefinition instanceof AnnotatedBeanDefinition) {
                MethodMetadata factoryMethodMetadata = ((AnnotatedBeanDefinition) beanDefinition).getFactoryMethodMetadata();
                if (factoryMethodMetadata != null) {
                    beanClassNames.add(factoryMethodMetadata.getReturnTypeName());
                }
            } else {
                String className = beanDefinition.getBeanClassName();
                if (className == null) {
                    className = beanFactory.getBean(beanDefinitionName).getClass().getName();
                }
                beanClassNames.add(className);
            }
        }
        for (String beanDefinitionName : beanFactory.getBeanDefinitionNames()) {
            BeanDefinitionRegistry beanRegistry = (BeanDefinitionRegistry) beanFactory;
            beanRegistry.removeBeanDefinition(beanDefinitionName);
        }

        writeToFile(beanClassNames);
    }

    private void writeToFile(ArrayList<String> beanClassNames) {
        try {
            File springBeansFile = new File("SpringBeans.txt");
            FileWriter fileWriter = new FileWriter(springBeansFile);

            Object[] distinctClassNames = beanClassNames.stream()
                    .distinct()
                    .toArray();
            Arrays.sort(distinctClassNames);

            for (Object beanClassName : distinctClassNames) {
                fileWriter.append(beanClassName.toString());
                fileWriter.append("\n");
            }

            fileWriter.flush();
            fileWriter.close();
        } catch (Throwable e) {
            System.out.println("Storing bean information failed");
        } finally {
            System.out.println("Finished post-processing bean factory in UtBot");
        }
    }

    @Override
    public int getOrder() {
        return PriorityOrdered.HIGHEST_PRECEDENCE;
    }
}

File: utbot-spring-analyzer/src/main/kotlin/post_processors/UtBotBeanFactoryPostProcessor.kt
package post_processors

import org.springframework.beans.factory.annotation.AnnotatedBeanDefinition
import org.springframework.beans.factory.config.BeanFactoryPostProcessor
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory
import org.springframework.beans.factory.support.BeanDefinitionRegistry
import org.springframework.core.PriorityOrdered

import java.io.File
import java.io.FileWriter
import java.util.Arrays

class UtBotBeanFactoryPostProcessor : BeanFactoryPostProcessor, PriorityOrdered {

    /**
     * Sets the priority of post processor to highest to avoid side effects from others.
     */
    override fun getOrder(): Int = PriorityOrdered.HIGHEST_PRECEDENCE

    override fun postProcessBeanFactory(beanFactory: ConfigurableListableBeanFactory) {
        println("Started post-processing bean factory in UtBot")

        val beanClassNames = findBeanClassNames(beanFactory)
        //TODO: will be replaced with more appropriate IPC approach.
        writeToFile(beanClassNames)

        // After desired post-processing is completed we destroy bean definitions
        // to avoid further possible actions with beans that may be unsafe.
        destroyBeanDefinitions(beanFactory)

        println("Finished post-processing bean factory in UtBot")
    }

    private fun findBeanClassNames(beanFactory: ConfigurableListableBeanFactory): ArrayList<String> {
        val beanClassNames = ArrayList<String>()
        for (beanDefinitionName in beanFactory.beanDefinitionNames) {
            val beanDefinition = beanFactory.getBeanDefinition(beanDefinitionName)

            if (beanDefinition is AnnotatedBeanDefinition) {
                val factoryMethodMetadata = beanDefinition.factoryMethodMetadata
                if (factoryMethodMetadata != null) {
                    beanClassNames.add(factoryMethodMetadata.returnTypeName)
                }
            } else {
                var className = beanDefinition.beanClassName
                if (className == null) {
                    className = beanFactory.getBean(beanDefinitionName).javaClass.name
                }
                className?.let { beanClassNames.add(it) }
            }
        }

        return beanClassNames
    }

    private fun destroyBeanDefinitions(beanFactory: ConfigurableListableBeanFactory) {
        for (beanDefinitionName in beanFactory.beanDefinitionNames) {
            val beanRegistry = beanFactory as BeanDefinitionRegistry
            beanRegistry.removeBeanDefinition(beanDefinitionName)
        }
    }

    private fun writeToFile(beanClassNames: ArrayList<String>) {
        try {
            val springBeansFile = File("SpringBeans.txt")
            val fileWriter = FileWriter(springBeansFile)

            val distinctClassNames = beanClassNames.stream()
                .distinct()
                .toArray()
            Arrays.sort(distinctClassNames)

            for (beanClassName in distinctClassNames) {
                fileWriter.append(beanClassName.toString())
                fileWriter.append("\n")
            }

            fileWriter.flush()
            fileWriter.close()

            println("Storing bean information completed successfully")
        } catch (e: Throwable) {
            println("Storing bean information failed with exception $e")
        }
    }
}